site/en/blog/renderingng-data-structures/index.md

```diff
@@ -0,0 +1,754 @@
+
+- Iterating in the depth-first search order is very fast.
+This is used very often because it is similar to caret movements.
+Since it's a flat list, depth-first search is just incrementing the array offset,
+providing fast iterations and memory locality.
+- It also provides breadth-first search, which is necessary when, for example,
+painting the background of line and inline boxes.
+- Knowing the number of descendants makes moving to the next sibling fast
+(just increment the array offset by that number).
+
+## Property trees
+
+As you know, the DOM is a tree of elements (plus text nodes),
+and CSS can apply various styles to elements.
+
+These come mostly in four flavors of effect:
+
+- **Layout:** inputs to the layout constraint algorithm.
+- **Paint:** how to paint and raster the element.
+(but not its descendants)
+- **Visual:** raster/draw effects applied to the DOM subtree,
+such as transforms, filters, and clipping
+- **Scrolling:** axis-aligned and rounded corner.
+clipping and scrolling of the contained subtree
+
+Property trees are data structures that explain how visual and scrolling effects apply to DOM elements.
+They provide the means to answer questions such as: where,
+relative to the screen, is a given DOM element,
+given its layout size and position?
+And: what sequence of GPU operations should be used to apply visual and scrolling effects?
+
+Visual and scrolling effects on the web are very complicated in their full glory.
+So the most important thing property trees do is translate that complexity
+into a single data structure that precisely represents their structure and meaning,
+while at the same time removing the rest of the complexity of the DOM and CSS.
+This lets us implement algorithms for compositing and scrolling with much more confidence:
+
+- Potentially error-prone geometry and other calculations
+can be centralized into one place.
+- The complexity of building and updating property trees
+is isolated into one rendering pipeline stage.
+- It's much easier and faster to send property trees around to different threads and processes than full DOM state,
+thereby making it possible to use them for many use cases.
+- The more use cases there are,
+the more wins we can get from geometry caching built on top,
+as they can reuse each others' caches.
+
+RenderingNG uses property trees for many purposes, including:
+
+- Separating compositing from paint, and compositing off the main thread.
+- Determining an optimal compositing / draw strategy.
+- Measuring
+[IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)
+geometry.
+- Avoiding work for offscreen elements and GPU texture tiles.
+- Efficient and accurate invalidation of paint and raster.
+- Measuring
+[layout shift](https://web.dev/cls/) and
+[largest contentful paint](https://web.dev/lcp/) in Core Web Vitals.
+
+Every web document has four separate property trees: transform, clip, effect and scroll.(*)
```

Comment:
I'm not massively keen either, but it's kind of a footnote to this section. The style guide isn't keen but suggests if you do it to use *. I think to put it further up would break up the flow of the section, so on balance, I'm probably less of a fan of that.